core/src/test/java/com/linecorp/armeria/client/ClientRequestContextTest.java

```diff
@@ -70,16 +86,214 @@ void mapCurrent() {
         assertThat(ClientRequestContext.mapCurrent(ctx -> "foo", () -> "bar")).isEqualTo("bar");
         assertThat(ClientRequestContext.mapCurrent(Function.identity(), null)).isNull();
 
-        final ClientRequestContext ctx = ClientRequestContext.of(req);
+        final ClientRequestContext ctx = clientRequestContext();
         try (SafeCloseable unused = ctx.push()) {
+            assertThat(onEnterExitStack.size()).isOne();
             assertThat(ClientRequestContext.mapCurrent(c -> "foo", () -> "bar")).isEqualTo("foo");
             assertThat(ClientRequestContext.mapCurrent(Function.identity(), null)).isSameAs(ctx);
         }
+        assertThat(onEnterExitStack.size()).isZero();
 
-        try (SafeCloseable unused = ServiceRequestContext.of(req).push()) {
+        try (SafeCloseable unused = serviceRequestContext().push()) {
             assertThatThrownBy(() -> ClientRequestContext.mapCurrent(c -> "foo", () -> "bar"))
                     .isInstanceOf(IllegalStateException.class)
                     .hasMessageContaining("not a client-side context");
         }
     }
+
+    @Test
+    void pushReentrance() {
+        final ClientRequestContext ctx = clientRequestContext();
+        try (SafeCloseable ignored = ctx.push()) {
+            assertThat(onEnterExitStack.size()).isOne();
+            try (SafeCloseable ignored2 = ctx.push()) {
+                assertThat(onEnterExitStack.size()).isOne();
```

Comment:
Could you add assertions that make sure `RequestContext.current()` returns the right context for all new test cases? e.g.

    assertThat(RequestContext.current()).isSameAs(ctx);